src/Microsoft.AspNetCore.Authentication.WsFederation/WsFederationHandler.cs

```diff
@@ -0,0 +1,347 @@
+        {
+            WsFederationMessage wsFederationMessage = null;
+
+            // assumption: if the ContentType is "application/x-www-form-urlencoded" it should be safe to read as it is small.
+            if (HttpMethods.IsPost(Request.Method)
+              && !string.IsNullOrEmpty(Request.ContentType)
+              // May have media/type; charset=utf-8, allow partial match.
+              && Request.ContentType.StartsWith("application/x-www-form-urlencoded", StringComparison.OrdinalIgnoreCase)
+              && Request.Body.CanRead)
+            {
+                var form = await Request.ReadFormAsync();
+    
+                wsFederationMessage = new WsFederationMessage(form.Select(pair => new KeyValuePair<string, string[]>(pair.Key, pair.Value)));
+            }
+
+            if (wsFederationMessage == null || !wsFederationMessage.IsSignInMessage)
+            {
+                if (Options.SkipUnrecognizedRequests)
+                {
+                    // Not for us?
+                    return HandleRequestResult.SkipHandler();
+                }
+
+                return HandleRequestResult.Fail("No message.");
+            }
+            
+            try
+            {
+                // Retrieve our cached redirect uri
+                var state = wsFederationMessage.Wctx;
+                // WsFed allows for uninitiated logins, state may be missing.
+                var properties = Options.StateDataFormat.Unprotect(state);
+
+                if (properties == null)
+                {
+                    properties = new AuthenticationProperties();
+                }
+                else
+                {
+                    // Extract the user state from properties and reset.
+                    properties.Items.TryGetValue(WsFederationDefaults.UserstatePropertiesKey, out var userState);
+                    wsFederationMessage.Wctx = userState;
+                }
+
+                var messageReceivedContext = new MessageReceivedContext(Context, Scheme, Options, properties)
+                {
+                    ProtocolMessage = wsFederationMessage
+                };
+                await Options.Events.MessageReceived(messageReceivedContext);
+                if (messageReceivedContext.Result != null)
+                {
+                    return messageReceivedContext.Result;
+                }
+
+                if (wsFederationMessage.Wresult == null)
+                {
+                    Logger.LogWarning("Received a sign-in message without a WResult.");
+                    return (HandleRequestResult)HandleRequestResult.NoResult();
+                }
+
+                var token = wsFederationMessage.GetToken();
+                if (string.IsNullOrEmpty(token))
+                {
+                    Logger.LogWarning("Received a sign-in message without a token.");
+                    return (HandleRequestResult)HandleRequestResult.NoResult();
```

Comment:
Same remark. Plus it's inconsistent with the OIDC handler.

Reply:
I don't see a direct equivalent in OIDC, it detects correct messages by looking for State and CorrelationId, both of which are optional for WsFed. That said, we should probably be using the SkipUnrecognizedRequests pattern here.

Reply:
> That said, we should probably be using the SkipUnrecognizedRequests pattern here.

It's something you're already doing at line 129 by making an `IsSignInMessage` check, which ensures that a `wa` parameter is present and corresponds to a sign-in operation. Once this check is made, you can consider the request as a valid WS-Fed message. Every missing parameter that is marked as required by the spec should result in an error.

Reply:
Fair.